src/adapter.jl

```diff
@@ -264,3 +264,16 @@ end
 # `false` if the default `hash` is used.
 Base.hash(::GapObj, h::UInt) = h
 Base.hash(::FFE, h::UInt) = h
+
+# iteration
+
+function Base.iterate(obj::GapObj)
+    Globals.IsList(obj) || Globals.IsCollection(obj) ||
+        throw(ArgumentError("object cannot be iterated"))
+    iterate(obj, (1, Globals.Length(obj)))
```

Comment:
That won't work for infinite obj, though, and even for finite ones can be very expensive as it forces a size computation. Granted, cases where this matter are more the exception than the norm.

Reply:
So yes, maybe we should define `Base.IteratorSize` to return `SizeUnknown()` (but I copied from Thomas' PR the iteration modality; I didn't know how to iterate otherwise).

Reply:
so for Thomas' PR, we know that we need all objects eventually, so in the infinite case, we'd have a problem anyway (we still might want to catch that explicitly a la `Globals.IsFinite(l) || error("cannot convert infinite collections")`